RHIDP-1923 - GKE: Document how RHDH can be installed in GKE (#715)
redhat-developer/red-hat-developers-documentation-rhdh: modules/installation/proc-rhdh-deploy-gke-helm.adoc

  enabled: false
upstream:
  service:
    type: NodePort
  ingress:
    enabled: true
    annotations:
      kubernetes.io/ingress.class: gce
      kubernetes.io/ingress.global-static-ip-name: <ADDRESS_NAME>
      networking.gke.io/managed-certificates: <rhdh_certificate_name>
      networking.gke.io/v1beta1.FrontendConfig: <ingress_security_config>
    className: gce
  backstage:
    image:
      pullSecrets:
      - rhdh-pull-secret
    podSecurityContext:
      fsGroup: 2000
  postgresql:
    image:
      pullSecrets:
      - rhdh-pull-secret
    primary:
      podSecurityContext:
        enabled: true
        fsGroup: 3000
  volumePermissions:
    enabled: true
----
. Run the following command in your terminal to deploy {product-short} using the latest version of Helm Chart and using the `values.yaml` file:
+
[source,terminal,subs="attributes+"]
----
helm -n <your_namespace> install -f values.yaml <your_deploy_name> \
  openshift-helm-charts/redhat-developer-hub \
  --version {product-chart-version}
----
+
For the latest Helm Chart version, see this https://github.com/openshift-helm-charts/charts/tree/main/charts/redhat/redhat/redhat-developer-hub[Helm Charts] repository.
+

> [NOTE]

Author: @hmanwani-rh The procedure is about _installing_ GKE, so is uninstalling or upgrading really part of the _verification_ of an _installation_ procedure.

I see your point though, so maybe we should have separate sections for upgrading and deleting your deployment after the installation procedure rather than placing them in the _Verfication_ section of the installation procedure?

> [NOTE]

Reviewer: I think we can certainly include a few commands in the verification section, as deleting and upgrading inherently confirm that something is already installed, which can then be deleted or upgraded. However, if there is substantial content to be added under the deleting or upgrading sections, having a separate procedure would make sense.

> [NOTE]

Author: @hmanwani-rh Moved delete and upgrade commands to Verification section.

* You have subscribed to `registry.redhat.io`. For more information, see https://access.redhat.com/RegistryAuthentication[{company-name} Container Registry Authentication].
* You have installed `kubectl`. For more information, see https://kubernetes.io/docs/tasks/tools/#kubectl[Install kubetl].
* You have installed the Google Cloud CLI. For more information, see https://cloud.google.com/sdk/docs/install[Install the gcloud CLI].
* You have logged in to your Google account and created a https://cloud.google.com/kubernetes-engine/docs/how-to/creating-an-autopilot-cluster[GKE Autopilot] or https://cloud.google.com/kubernetes-engine/docs/how-to/creating-a-zonal-cluster[GKE Standard] cluster.
* You have installed Helm 3 or the latest version. For more information, see https://docs.aws.amazon.com/eks/latest/userguide/helm.html[Using Helm with Amazon {eks-short}].

* You have configured a domain name for your {product-short} instance.
* You have reserved a static external Premium IPv4 Global IP address that is not attached to any VM. For more information see https://cloud.google.com/vpc/docs/reserve-static-external-ip-address#reserve_new_static[Reserve a new static external IP address]

> * You have configured the DNS records for your domain name to point to the IP address that have reseved. 

Reviewer: ```suggestion
* You have configured the DNS records for your domain name to point to the IP address that has been reserved. 
```

+

> [NOTE]

Reviewer: ```suggestion
[NOTE]
====
```

> You need to create an `A` record with the value equal to the IP address. This can take up to one hour to propagate.

Reviewer: ```suggestion
You need to create an `A` record with the value equal to the IP address. This can take up to one hour to propagate.
====
```

//* You have an entry in the AWS Certificate Manager (ACM) for your preferred domain name. Make sure to keep a record of your Certificate ARN.
//* You have set the context to the {eks-short} cluster in your current `kubeconfig`. For more information, see https://docs.aws.amazon.com/eks/latest/userguide/create-kubeconfig.html[Creating or updating a kubeconfig file for an Amazon {eks-short} cluster].
//* You have installed Helm 3 or the latest. For more information, see https://docs.aws.amazon.com/eks/latest/userguide/helm.html[Using Helm with Amazon {eks-short}].

.Procedure

. Go to your terminal and run the following command to add the Helm chart repository containing the {product-short} chart to your local Helm registry:
+
--
[source,terminal]
----
helm repo add openshift-helm-charts https://charts.openshift.io/
----
--

. Create a pull secret using the following command:
+
--

> [source,terminal]

Reviewer: How about we add a caption to all the fragments. For example:
```suggestion
.Example command to create a pull secret
[source,terminal]
```

Author: Captions added to fragments.

----
kubectl -n <your-namespace> create secret docker-registry rhdh-pull-secret \ <1>
    --docker-server=registry.redhat.io \
    --docker-username=<user_name> \ <2>
    --docker-password=<password> \ <3>
    --docker-email=<email> <4>
----
<1> Enter your {gke-short} namespace in the command.
<2> Enter your username in the command.
<3> Enter your password in the command.
<4> Enter your email address in the command.

The created pull secret is used to pull the {product-short} images from the {company-name} Ecosystem.
--

> . Set up a Google-managed certificate by creating a `ManagedCertificate` object that you will attach to the Ingress.

Reviewer: ```suggestion
. Set up a Google-managed certificate by creating a `ManagedCertificate` object that you must attach to the ingress.
```

+
--
[source,yaml,subs="attributes+"]
----
apiVersion: networking.gke.io/v1
kind: ManagedCertificate
metadata:
  name: <rhdh_certificate_name>
spec:
  domains:
    - <rhdh_domain_name>
----
--
For more information about setting up a Google-managed certificate, see https://cloud.google.com/kubernetes-engine/docs/how-to/managed-certs?hl=en#setting_up_a_google-managed_certificate[Setting up a Google-managed certificate].

> . Create a `FrontendConfig` object to set a policy for redirecting to HTTPS. You will attach this policy to the Ingress. 

Reviewer: ```suggestion
. Create a `FrontendConfig` object to set a policy for redirecting to HTTPS. You must attach this policy to the ingress. 
```

+
--
[source,yaml,subs="attributes+"]
----
apiVersion: networking.gke.io/v1beta1
kind: FrontendConfig
metadata:
  name: <ingress_security_config>
spec:
  sslPolicy: gke-ingress-ssl-policy-https
  redirectToHttps:
    enabled: true
----
--
For more information about setting a policy to redirect to HTTPS, see https://cloud.google.com/kubernetes-engine/docs/how-to/ingress-configuration?hl=en#https_redirect[HTTP to HTTPS redirects].

. Create a file named `values.yaml` using the following template:
+

> [source,yaml,subs="attributes+"]

Reviewer: ```suggestion
.Example `values.yaml` file
[source,yaml,subs="attributes+"]
```

----
global:
  host: <rhdh_domain_name>
route:
  enabled: false
upstream:
  service:
    type: NodePort
  ingress:
    enabled: true
    annotations:
      kubernetes.io/ingress.class: gce
      kubernetes.io/ingress.global-static-ip-name: <ADDRESS_NAME>
      networking.gke.io/managed-certificates: <rhdh_certificate_name>
      networking.gke.io/v1beta1.FrontendConfig: <ingress_security_config>
    className: gce
  backstage:
    image:
      pullSecrets:
      - rhdh-pull-secret
    podSecurityContext:
      fsGroup: 2000
  postgresql:
    image:
      pullSecrets:
      - rhdh-pull-secret
    primary:
      podSecurityContext:
        enabled: true
        fsGroup: 3000
  volumePermissions:
    enabled: true
----

> . Run the following command in your terminal to deploy {product-short} using the latest version of Helm Chart and using the values.yaml file created in the previous step:

Reviewer: ```suggestion
. Run the following command in your terminal to deploy {product-short} using the latest version of Helm Chart and using the `values.yaml` file:
```

+
[source,terminal,subs="attributes+"]
----
helm -n <your_namespace> install -f values.yaml <your_deploy_name> \
  openshift-helm-charts/redhat-developer-hub \
  --version {product-chart-version}
----
+

> For the latest Helm Chart version, see https://github.com/openshift-helm-charts/charts/tree/main/charts/redhat/redhat/redhat-developer-hub

Reviewer: This link is missing the text.

+

> . Confirm that the deployment is complete.

Reviewer: I think this and the following step would go in the Verification section.

+
[source,terminal,subs="attributes+"]
----
kubectl get deploy <you_deploy_name>-developer-hub -n <your_namespace>
----

. Verify that the service and ingress were created.
+
[source,terminal,subs="attributes+"]
----
kubectl get service -n <your_namespace>
kubectl get ingress -n <your_namespace>
----
+
[NOTE]
Wait for the `ManagedCertificate` to be provisioned. This can take a couple of hours.

> . Access RHDH with `https://<rhdh_domain_name>`

Reviewer: ```suggestion
. Access {product-very-short} with `https://<rhdh_domain_name>`
```

When you install the {product-short} Helm chart in {gke-brand-name} ({gke-short}), it orchestrates the deployment of a {product-short} instance, which provides a robust developer platform within the {gke-short} ecosystem.

.Prerequisites
* You have subscribed to `registry.redhat.io`. For more information, see https://access.redhat.com/RegistryAuthentication[{company-name} Container Registry Authentication].
* You have installed `kubectl`. For more information, see https://kubernetes.io/docs/tasks/tools/#kubectl[Install kubetl].
* You have installed the Google Cloud CLI. For more information, see https://cloud.google.com/sdk/docs/install[Install the gcloud CLI].
* You have logged in to your Google account and created a https://https://cloud.google.com/kubernetes-engine/docs/how-to/creating-an-autopilot-cluster[GKE Autopilot] or https://cloud.google.com/kubernetes-engine/docs/how-to/creating-a-zonal-cluster[GKE Standard] cluster.
* You have installed Helm 3 or the latest version. For more information, see https://docs.aws.amazon.com/eks/latest/userguide/helm.html[Using Helm with Amazon {eks-short}].

* You have configured a domain name for your {product-short} instance.

> * You have reserved a static external Premium IPv4 Global IP address that is not attached to any VM.
> * You have configured the DNS records for your domain name to point to the IP address that have reseved. For more information see https://cloud.google.com/vpc/docs/reserve-static-external-ip-address#reserve_new_static[Reserve a new static external IP address]

Reviewer 2: The link is related to the IP reservation
```suggestion
* You have reserved a static external Premium IPv4 Global IP address that is not attached to any VM. For more information see https://cloud.google.com/vpc/docs/reserve-static-external-ip-address#reserve_new_static[Reserve a new static external IP address]
* You have configured the DNS records for your domain name to point to the IP address that have reseved. 
```

  postgresql:
    image:
      pullSecrets:
      - rhdh-pull-secret
    primary:
      podSecurityContext:
        enabled: true
        fsGroup: 3000
  volumePermissions:
    enabled: true
----
. Run the following command in your terminal to deploy {product-short} using the latest version of Helm Chart and using the values.yaml file created in the previous step:
+
[source,terminal,subs="attributes+"]
----
helm -n <your_namespace> install -f values.yaml <your_deploy_name> \
  openshift-helm-charts/redhat-developer-hub \
  --version {product-chart-version}
----
+
For the latest Helm Chart version, see https://github.com/openshift-helm-charts/charts/tree/main/charts/redhat/redhat/redhat-developer-hub
+
. Confirm that the deployment is complete.
+
[source,terminal,subs="attributes+"]
----
kubectl get deploy <you_deploy_name>-developer-hub -n <your_namespace>
----

. Verify that the service and ingress were created.
+
[source,terminal,subs="attributes+"]
----
kubectl get service -n <your_namespace>
kubectl get ingress -n <your_namespace>
----
+
[NOTE]
Wait for the `ManagedCertificate` to be provisioned. This can take a couple of hours.

. Access RHDH with `https://<rhdh_domain_name>`

> +
> [IMPORTANT]
> Use the HTTPS protocol, not HTTP.

Reviewer 2: We actually don't need this note for GKE, because we are setting a redirect from HTTP to HTTPS. My bad - I forgot this in the draft.
```suggestion
```
